sdk/ai/azure-ai-projects/azure/ai/projects/models/_patch.py

```diff
@@ -913,119 +1109,319 @@ async def execute_tool_calls(self, tool_calls: List[Any]) -> Any:
+                self.on_done()
+                self.done = True  # Mark the stream as done
+            else:
+                self.on_unhandled_event(event_type, event_data_obj)
+        except Exception as e:  # pylint: disable=broad-exception-caught
+            logging.error("Error in event handler for event '%s': %s", event_type, e)
+        return event_type, event_data_obj, self.custom_data
+
+    def on_message_delta(self, delta: "MessageDeltaChunk") -> None:
         """Handle message delta events.
 
         :param MessageDeltaChunk delta: The message delta.
         """
 
-    async def on_thread_message(self, message: "ThreadMessage") -> None:
+    def on_thread_message(self, message: "ThreadMessage") -> None:
         """Handle thread message events.
 
         :param ThreadMessage message: The thread message.
         """
 
-    async def on_thread_run(self, run: "ThreadRun") -> None:
+    def on_thread_run(self, run: "ThreadRun") -> None:
         """Handle thread run events.
 
         :param ThreadRun run: The thread run.
         """
 
-    async def on_run_step(self, step: "RunStep") -> None:
+    def on_run_step(self, step: "RunStep") -> None:
         """Handle run step events.
 
         :param RunStep step: The run step.
         """
 
-    async def on_run_step_delta(self, delta: "RunStepDeltaChunk") -> None:
+    def on_run_step_delta(self, delta: "RunStepDeltaChunk") -> None:
         """Handle run step delta events.
 
         :param RunStepDeltaChunk delta: The run step delta.
         """
 
-    async def on_error(self, data: str) -> None:
+    def on_error(self, data: str) -> None:
         """Handle error events.
 
         :param str data: The error event's data.
         """
 
-    async def on_done(self) -> None:
+    def on_done(self) -> None:
         """Handle the completion of the stream."""
 
-    async def on_unhandled_event(self, event_type: str, event_data: Any) -> None:
+    def on_unhandled_event(self, event_type: str, event_data: str) -> None:
         """Handle any unhandled event types.
 
         :param str event_type: The event type.
         :param Any event_data: The event's data.
         """
 
 
-class AsyncAgentRunStream(AsyncIterator[Tuple[str, Union[str, StreamEventData]]]):
+class AsyncAgentRunStream(Generic[BaseAsyncAgentEventHandlerT]):
     def __init__(
         self,
         response_iterator: AsyncIterator[bytes],
-        submit_tool_outputs: Callable[[ThreadRun, Optional[AsyncAgentEventHandler]], Awaitable[None]],
-        event_handler: Optional["AsyncAgentEventHandler"] = None,
+        submit_tool_outputs: Callable[[ThreadRun, BaseAsyncAgentEventHandlerT], Awaitable[None]],
+        event_handler: BaseAsyncAgentEventHandlerT,
     ):
         self.response_iterator = response_iterator
         self.event_handler = event_handler
         self.done = False
         self.buffer = ""
         self.submit_tool_outputs = submit_tool_outputs
+        self.event_handler._init(
```

Comment:
We should not call __init__ of an instance passed it/it would be very unexpected. Is the intent to provide the **type** of the event handler to use and not the instance?

Reply:
Case 1: event_handler object is instantiated by users and passed to create_stream.   
Case 2: If users don't provide an event handler object in create_stream, create_stream create an object for them.
I probably can do what you said for case 2.   But for case 1, there is no streaming happening while users define the event_handler.   So I need to offer _init as a separate function to fill in the event_handler.

Reply:
I talked to Johan, the function name looks confusing as it looks liked a constructor.   So I make a ticket to follow up post release.
https://github.com/orgs/ai-platform-ml-platform/projects/7/views/1?pane=issue&itemId=90611930